cms/src/builder.rs

```diff
@@ -705,32 +704,104 @@ impl RecipientInfoBuilder for KekRecipientInfoBuilder {
     /// token such as a smart card.
     pub key_derivation_alg: Option<AlgorithmIdentifierOwned>,
     /// Encryption algorithm to be used for key encryption
     pub key_enc_alg: AlgorithmIdentifierOwned,
+    /// Provided password encryptor
+    pub key_encryptor: P,
+    /// Random number generator
+    pub rng: &'r mut R,
 }
 
-impl PasswordRecipientInfoBuilder {
+impl<'r, P, R> PasswordRecipientInfoBuilder<'r, P, R>
+where
+    P: PwriEncryptor,
+    R: CryptoRngCore,
+{
     /// Creates a `PasswordRecipientInfoBuilder`
-    pub fn new(
-        key_derivation_alg: Option<AlgorithmIdentifierOwned>,
-        key_enc_alg: AlgorithmIdentifierOwned,
-    ) -> Result<PasswordRecipientInfoBuilder> {
+    /// `key_derivation_alg`: (optional) Algorithm used to derive the
+    ///     key-encryption key from the shared secret (password)
+    /// `key_enc_alg`: Algorithm used to (symmetrically) encrypt the
+    ///     content-encryption key
+    /// `key_encryptor`: Provided encryptor, which is used to encrypt
+    ///     the content-encryption key
+    /// `rng`: Random number generator, required for padding values.
+    pub fn new(key_encryptor: P, rng: &'r mut R) -> Result<PasswordRecipientInfoBuilder<'r, P, R>> {
         Ok(PasswordRecipientInfoBuilder {
-            key_derivation_alg,
-            key_enc_alg,
+            key_derivation_alg: key_encryptor.key_derivation_algorithm()?,
+            key_enc_alg: key_encryptor.key_encryption_algorithm()?,
+            key_encryptor,
+            rng,
         })
     }
+
+    /// Wrap the content-encryption key according to [RFC 3211, §2.3.1]:
+    ///     ....
+    ///     The formatted CEK block then looks as follows:
+    ///     CEK byte count || check value || CEK || padding (if required)
+    ///
+    /// [RFC 3211, §2.3.1]: https://www.rfc-editor.org/rfc/rfc3211#section-2.3.1
+    fn wrap_content_encryption_key(&mut self, content_encryption_key: &[u8]) -> Result<Vec<u8>> {
+        let content_encryption_key_length = content_encryption_key.len();
+        let wrapped_key_length_wo_padding = 1 + 3 + content_encryption_key_length;
+        let key_enc_alg_blocklength = P::BLOCK_LENGTH;
+        let padding_length = if wrapped_key_length_wo_padding < 2 * key_enc_alg_blocklength {
+            2 * key_enc_alg_blocklength - wrapped_key_length_wo_padding
+        } else {
+            0
+        };
+
+        let cek_byte_count: u8 = content_encryption_key.len().try_into().map_err(|_| {
+            Error::Builder("Content encryption key length must not exceed 255".to_string())
+        })?;
+        let mut wrapped_cek: Vec<u8> = Vec::with_capacity(4 + padding_length);
+        wrapped_cek.push(cek_byte_count);
+        wrapped_cek.push(0xff ^ content_encryption_key[0]);
+        wrapped_cek.push(0xff ^ content_encryption_key[1]);
+        wrapped_cek.push(0xff ^ content_encryption_key[2]);
+        if padding_length > 0 {
```

Comment:
```suggestion
        wrapped_cek.push(0xff ^ content_encryption_key[2]);
        wrapped_cek.append(content_encryption_key);
        if padding_length > 0 {
```

We forgot the CEK here right?

Reply:
That could explain, why decryption failed last night :smiley:

Reply:
hu, sorry I always get mixed up with the vec methods.